src/ray/util/shared_lru.h

```diff
@@ -27,32 +27,38 @@
 // // Check and consume `val`.
 //
 // TODO(hjiang):
-// 1. Add template arguments for key hash and key equal, to pass into absl::flat_hash_map.
-// 2. Provide a key hash wrapper to save a copy.
-// 3. flat hash map supports heterogeneous lookup, expose `KeyLike` templated interface.
-// 4. Add a `GetOrCreate` interface, which takes factory function to creation value.
-// 5. For thread-safe cache, add a sharded container wrapper to reduce lock contention.
+// 1. Write a wrapper around KeyHash and KeyEq, which takes std::reference_wrapper<Key>,
+// so we could store keys only in std::list, and reference in absl::flat_hash_map.
+// 2. Add a `GetOrCreate` interface, which takes factory function to creation value.
+// 3. For thread-safe cache, add a sharded container wrapper to reduce lock contention.
 
 #pragma once
 
 #include <cstdint>
 #include <list>
 #include <memory>
 #include <mutex>
-#include <optional>
 #include <string>
 #include <utility>
 
 #include "absl/container/flat_hash_map.h"
+#include "absl/container/internal/hash_function_defaults.h"
 #include "src/ray/util/logging.h"
 
 namespace ray::utils::container {
 
-template <typename Key, typename Val>
+// TODO(hjiang): `absl::container_internal::hash_default_` has been promoted to stable
+// public interface in the latest abseil, update after we bump up abseil version.
+template <typename Key,
+          typename Val,
+          typename KeyHash = absl::container_internal::hash_default_hash<Key>,
```

Comment:
Q: Even if we don't provide KeyHash, KeyEq in our cache template arguments, the inner absl map would still use them, and hence the templated Get and Delete should automatically work. Why do we need to provide KeyHash, KeyEq arguments?

Reply:
Sorry I don't understand what you mean, could you please maybe provide a code snippet what you're thinking?

Reply:
No, without customized KeyHash and KeyEq, you cannot use self-defined struct as key.

Reply:
I think you might have some mis-understanding for this PR, I'm trying to do two things:
- Add heterogeneous lookup and delete support, via exposing `KeyLike` template interface
- Add customized hash and equal function, so you could use self-defined class as key

These are two completely unrelated changes. To test these changes,
- `TestClass` is used in test cases for item (2)
- Existing lookup and delete functions are updated to leverage heterogeneous access

Reply:
I think maybe it's better to split the PR into two parts for easier code review next time.